comms/src/common/rate_limit.rs

```diff
@@ -0,0 +1,172 @@
+//! a given time interval. The underlying stream will begin to buffer and produce backpressure if producers exceed the
+//! capacity and restock_intervals.
+
+// This is slightly changed from the libra rate limiter implementation
+
+use futures::{stream::Fuse, FutureExt, Stream, StreamExt};
+use pin_project::pin_project;
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::{
+    sync::{OwnedSemaphorePermit, Semaphore},
+    time,
+    time::Interval,
+};
+
+pub trait RateLimit: Stream {
+    /// Consumes the stream and returns a rate-limited stream that only polls the underlying stream
+    /// a maximum of `capacity` times within `restock_interval`.
+    fn rate_limit(self, capacity: usize, restock_interval: Duration) -> RateLimiter<Self>
+    where Self: Sized {
+        RateLimiter::new(self, capacity, restock_interval)
+    }
+}
+
+impl<T: Stream> RateLimit for T {}
+
+#[pin_project]
+#[must_use = "streams do nothing unless polled"]
+pub struct RateLimiter<T> {
+    /// The inner stream to poll when a permit has been acquired
+    #[pin]
+    stream: T,
+    /// An interval stream that "restocks" the permits
+    #[pin]
+    interval: Fuse<Interval>,
+    /// The maximum permits to issue
+    capacity: usize,
+    /// A semaphore that holds the permits
+    permits: Arc<Semaphore>,
+    permit_future: Option<Pin<Box<dyn Future<Output = OwnedSemaphorePermit> + Send>>>,
+    permit_acquired: bool,
+}
+
+impl<T: Stream> RateLimiter<T> {
+    pub fn new(stream: T, capacity: usize, restock_interval: Duration) -> Self {
+        RateLimiter {
+            stream,
+            capacity,
+
+            interval: time::interval(restock_interval).fuse(),
+            // `interval` starts immediately, so we can start with zero permits
+            permits: Arc::new(Semaphore::new(0)),
+            permit_future: None,
+            permit_acquired: false,
+        }
+    }
+}
+
+impl<T: Stream> Stream for RateLimiter<T> {
+    type Item = T::Item;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // "Restock" permits once interval is ready
+        if let Poll::Ready(Some(_)) = self.as_mut().project().interval.poll_next(cx) {
+            self.permits
+                .add_permits(self.capacity - self.permits.available_permits());
+        }
+
+        // Attempt to acquire a permit
+        if !self.permit_acquired {
+            // Set the permit future
+            if self.permit_future.is_none() {
+                let permits = self.permits.clone();
+                *self.as_mut().project().permit_future = Some(permits.acquire_owned().boxed());
+            }
+
+            // Wait until a permit is acquired
+            let permit = futures::ready!(self
+                .as_mut()
+                .project()
+                .permit_future
+                .as_mut()
+                .unwrap()
```

Comment:
Perhaps an expect so that we can see a message if this breaks?

Reply:
Given that the code above sets this (checks if it's none and sets it if it isnt) it's not possible to break - maybe a comment to clarify that point